src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Windows.cs

```diff
@@ -284,40 +367,64 @@ private void InnerStartOperationReceive()
             //   An array of WSABuffer descriptors is allocated.
         }
 
-        internal unsafe SocketError DoOperationReceive(SafeCloseSocket handle, out SocketFlags flags)
+        internal unsafe SocketError DoOperationReceive(SafeCloseSocket handle)
         {
             PrepareIOCPOperation();
 
-            flags = _socketFlags;
-
+            int result;
             int bytesTransferred;
-            SocketError socketError;
+            SocketFlags flags = _socketFlags;
             if (_buffer != null)
             {
                 // Single buffer case.
-                socketError = Interop.Winsock.WSARecv(
+                WSABuffer wsaBuffer = new WSABuffer(GetPinnedSingleBufferPointer(), _count);
+                result = Interop.Winsock.WSARecv(
                     handle,
-                    ref _wsaBuffer,
+                    &wsaBuffer,
                     1,
-                    out bytesTransferred,
-                    ref flags,
+                    &bytesTransferred,
+                    &flags,
                     _ptrNativeOverlapped,
-                    IntPtr.Zero);
+                    null);
             }
             else
             {
                 // Multi buffer case.
-                socketError = Interop.Winsock.WSARecv(
-                    handle,
-                    _wsaBufferArray,
-                    _wsaBufferArray.Length,
-                    out bytesTransferred,
-                    ref flags,
-                    _ptrNativeOverlapped,
-                    IntPtr.Zero);
+                int bufferCount = _bufferListInternal.Count;
+                if (bufferCount <= WSABuffer.StackAllocLimit)
+                {
+                    WSABuffer* wsaBuffers = stackalloc WSABuffer[bufferCount];
+                    PopulateWSABuffersArray(wsaBuffers);
+
+                    result = Interop.Winsock.WSARecv(
+                        handle,
+                        wsaBuffers,
+                        bufferCount,
+                        &bytesTransferred,
+                        &flags,
+                        _ptrNativeOverlapped,
+                        null);
+                }
+                else
+                {
+                    WSABuffer[] wsaHeapBuffers = new WSABuffer[bufferCount];
+                    fixed (WSABuffer* wsaBuffers = wsaHeapBuffers)
```

Comment:
Nit: as a small perf improvement, this could be:
```C#
fixed (WSABuffer* wsaBuffers = &wsaHeapBuffers[0])
```
instead of:
```C#
fixed (WSABuffer* wsaBuffers = wsaHeapBuffers)
```
The latter results in the C# compiler emitting checks to ensure that wsaHeapBuffers isn't null and isn't empty (in both cases it'll emit code to set wsaBuffers to null).  Since you know here that it's neither of those, you can shrink the IL by explicitly getting a pointer to the 0th element.  Same logic would apply anywhere else you're sure the array being fixed is non-empty.